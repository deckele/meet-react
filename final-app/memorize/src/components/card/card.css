.card {
	height: 3.7rem;
	width: 3.7rem;
	margin: 0.2rem 0.2rem;
	background: #222;;
	font-size: 0;
	font-style: normal;
	color: #ffffff;
	border-radius: 5px;
	cursor: pointer;
	display: flex;
	justify-content: center;
	align-items: center;
	box-shadow: 5px 2px 20px 0 rgba(46, 61, 73, 0.5);
}

.card.open {
	transform: rotateY(0);
	background: #02b3e4;
	cursor: default;
	animation-name: flipInY;
	-webkit-backface-visibility: visible !important;
	backface-visibility: visible !important;
	animation-duration: .75s;
}

.card.show {
	font-size: 33px;
    background: #000;
    border: 4px solid #999;
}

.card.match {
	cursor: default;
	background: #E5F720;
	font-size: 33px;
	animation-name: rubberBand;
	-webkit-backface-visibility: visible !important;
	backface-visibility: visible !important;
	animation-duration: .75s;
}

.card.unmatched {
	animation-name: pulse;
	-webkit-backface-visibility: visible !important;
	backface-visibility: visible !important;
	animation-duration: .75s;
	background: #e2043b;
}

.card.disabled {
	pointer-events: none;
	opacity: 0.9;
}

/****** Media queries
***************************/
@media (max-width: 320px) {
	.card {
		height: 4.7rem;
		width: 4.7rem;
	}
}


/* For Tablets and larger screens
****************/
@media (min-width: 768px) {
	.card {
		height: 160px;
		width: 125px;
	}
}

