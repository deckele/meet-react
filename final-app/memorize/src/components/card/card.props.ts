import { CardType } from '../../models/card-types';

export interface CardProps {
    isOpen: boolean;
    isMatched: boolean;
    disabled: boolean;
    type: CardType;
    onClick(): void;
}
