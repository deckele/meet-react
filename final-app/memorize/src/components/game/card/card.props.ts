export interface CardProps {
    isOpen: boolean;
    disabled: boolean;
    value: string;
    isMatched: boolean;
    onClick(): void;
}
