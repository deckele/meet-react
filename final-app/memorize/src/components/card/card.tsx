import * as React from 'react';
import { CardProps } from './card.props';
import './card.css';

export const Card = ({type, onClick, isFlipped}: CardProps) =>
                            (<i className={getClasses(true, false)} onClick={() => onClick()}>{type}</i>);

function getClasses(isFlipped: boolean, isMatched: boolean) {
    let className = 'card';
    if (isFlipped) {
        className += ' show';
    }
    if (isMatched) {
        className += ' disabeld';
    }
    return className;
}