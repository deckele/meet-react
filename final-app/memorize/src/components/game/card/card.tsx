import * as React from 'react';
import { CardProps } from './card.props';
import './card.css';

export class Card extends React.Component<CardProps> {
    constructor(props: CardProps) {
        super(props);
        this.selectCard = this.selectCard.bind(this);
    }

    render() {
        return (
            <i className={this.cardClasses} onClick={this.selectCard}>
                {this.props.value}
            </i>
        );
    }

    private get cardClasses() {
        let className = 'card';
        if (this.props.isOpen) {
            className += ' open show disabled';
        }
        if (this.props.isMatched) {
            className += ' match';
        }
        if (this.props.disabled) {
            className += ' disabled';
        }
        return className;
    }

    private selectCard() {
        if (this.props.isMatched || this.props.disabled) {
            return;
        }
        this.props.onClick();
    }
}