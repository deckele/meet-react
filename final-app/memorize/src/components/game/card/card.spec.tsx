import * as React from 'react';
import { shallow } from 'enzyme';
import { Card } from './card';

describe('Button', () => {
    it('should render button correctly', () => {
        const wrapper = shallow(
            <Card
                onClick={() => {/**/}}
                disabled={false}
                isMatched={false}
                isOpen={false}
                value="A"
            />
        );
        expect(wrapper).toMatchSnapshot();
    });
    it('should not call onClick when clicking the card and card is disabled', () => {
        const onClick = jest.fn();
        const wrapper = shallow(
            <Card
                onClick={onClick}
                disabled={true}
                isMatched={false}
                isOpen={false}
                value="A"
            />
        );
        wrapper.find('i').simulate('click');
        expect(onClick).not.toHaveBeenCalled();
    });
    it('should not call onClick when clicking the card and card is matched', () => {
        const onClick = jest.fn();
        const wrapper = shallow(
            <Card
                onClick={() => {/**/}}
                disabled={false}
                isMatched={true}
                isOpen={false}
                value="A"
            />
        );
        wrapper.find('i').simulate('click');
        expect(onClick).not.toHaveBeenCalled();
    });
    it('should call onClick when clicking the card', () => {
        const onClick = jest.fn();
        const wrapper = shallow(
            <Card
                onClick={() => {/**/}}
                disabled={false}
                isMatched={false}
                isOpen={true}
                value="A"
            />
        );
        wrapper.find('i').simulate('click');
        expect(onClick).toHaveBeenCalled();
    });
    it('should display card as open and disabled when isOpen prop is true', () => {
        const wrapper = shallow(
            <Card
                onClick={() => {/**/}}
                disabled={false}
                isMatched={false}
                isOpen={true}
                value="A"
            />
        );
        const className = wrapper.find('i').prop('className');
        expect(className).toContain('open');
        expect(className).toContain('disabled');
        expect(className).toContain('show');
    });

    it('should display card as matched when isMatched prop is true', () => {
        const wrapper = shallow(
            <Card
                onClick={() => {/**/}}
                disabled={false}
                isMatched={true}
                isOpen={false}
                value="A"
            />
        );
        const className = wrapper.find('i').prop('className');
        expect(className).toContain('match');
    });
});