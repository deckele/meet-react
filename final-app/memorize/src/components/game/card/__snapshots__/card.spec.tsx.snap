// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render button correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Card
    disabled={false}
    isMatched={false}
    isOpen={false}
    onClick={[Function]}
    value="A"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": "A",
      "className": "card",
      "onClick": [Function],
    },
    "ref": null,
    "rendered": "A",
    "type": "i",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "A",
        "className": "card",
        "onClick": [Function],
      },
      "ref": null,
      "rendered": "A",
      "type": "i",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
